astropy/modeling/core.py

```diff
@@ -1871,32 +1871,46 @@ def __reduce__(cls):
 
     @property
     def submodel_names(cls):
-        if cls._submodel_names is not None:
-            return cls._submodel_names
-
-        by_name = defaultdict(list)
-
-        for idx, submodel in enumerate(cls._get_submodels()):
-            # Keep track of the original sort order of the submodels
-            by_name[submodel.name].append(idx)
-
-        names = []
-        for basename, indices in six.iteritems(by_name):
-            if len(indices) == 1:
-                # There is only one model with this name, so it doesn't need an
-                # index appended to its name
-                names.append((basename, indices[0]))
-            else:
-                for idx in indices:
-                    names.append(('{0}_{1}'.format(basename, idx), idx))
-
-        # Sort according to the models' original sort orders
-        names.sort(key=lambda k: k[1])
+        # If results are not cached, then set the cache
+        if cls._submodel_names is None:
+            
+            # Duplicate names must be disambiguated.
+            seen = {}
+            names = []
+            for idx, submodel in enumerate(cls._get_submodels()):
+                name = submodel.name
+                if name is None:
+                    # No need to disambiguate None
+                    names.append(name)
```

Comment:
I will just copy below my comment from the previous PR since this is a change in behavior. It's also the cause for the Travis failure.

Thinking a bit more about changing how names are created for models without user set names - in master the name is set to a string of "None_#", where # indicates the submodel number in the adopted ordering. This PR changes it to None. I prefer the naming in master for two reasons. Names are consistently strings so `submodel_names` returns a tuple of strings. The ordering is clear and although there's no user friendly name, the # shows how to get a model by indexing the compound model. It's not perfect but is still a bit more useful than just None.